File: db4oj/core/src/com/db4o/cs/messages/MCreateClass.java
/* Copyright (C) 2004   db4objects Inc.   http://www.db4o.com */

package com.db4o.cs.messages;

import com.db4o.*;
import com.db4o.cs.*;
import com.db4o.reflect.ReflectClass;

public final class MCreateClass extends MsgD {

	public final boolean processAtServer(YapServerThread serverThread) {
        YapStream stream = stream();
        Transaction trans = stream.getSystemTransaction();
        YapWriter returnBytes = new YapWriter(trans, 0);
        try{
            ReflectClass claxx = trans.reflector().forName(readString());
            if (claxx != null) {
                synchronized (streamLock()) {
                    try {
                        YapClass yapClass = stream.produceYapClass(claxx);
                        if (yapClass != null) {
                            stream.checkStillToSet();
                            yapClass.setStateDirty();
                            yapClass.write(trans);
                            trans.commit();
                            returnBytes = stream.readWriterByID(trans, yapClass.getID());
                            serverThread.write(Msg.OBJECT_TO_CLIENT.getWriter(returnBytes));
                            return true;
    
                        } 
                        
                        // TODO: handling, if the class can't be created
                        
                    } catch (Throwable t) {
                        if (Deploy.debug) {
                            System.out.println("MCreateClass failed");
                        }
                    }
                }
            }
        }catch(Throwable th){
            if (Deploy.debug) {
                System.out.println("MCreateClass failed");
            }
        }
        serverThread.write(Msg.FAILED);
        return true;
    }
}

File: db4on/core/src/core/com/db4o/internal/cs/messages/MCreateClass.cs
namespace com.db4o.@internal.cs.messages
{
	public sealed class MCreateClass : com.db4o.@internal.cs.messages.MsgD
	{
		public sealed override bool ProcessAtServer(com.db4o.@internal.cs.ServerMessageDispatcher
			 serverThread)
		{
			com.db4o.@internal.ObjectContainerBase stream = Stream();
			com.db4o.@internal.Transaction trans = stream.GetSystemTransaction();
			com.db4o.reflect.ReflectClass claxx = trans.Reflector().ForName(ReadString());
			if (claxx == null)
			{
				return WriteFailedMessage(serverThread);
			}
			lock (StreamLock())
			{
				try
				{
					com.db4o.@internal.ClassMetadata yapClass = stream.ProduceYapClass(claxx);
					if (yapClass == null)
					{
						return WriteFailedMessage(serverThread);
					}
					stream.CheckStillToSet();
					yapClass.SetStateDirty();
					yapClass.Write(trans);
					trans.Commit();
					com.db4o.@internal.StatefulBuffer returnBytes = stream.ReadWriterByID(trans, yapClass
						.GetID());
					com.db4o.@internal.cs.messages.MsgD createdClass = com.db4o.@internal.cs.messages.Msg
						.OBJECT_TO_CLIENT.GetWriter(returnBytes);
					serverThread.Write(createdClass);
				}
				catch (com.db4o.foundation.Db4oRuntimeException)
				{
					WriteFailedMessage(serverThread);
				}
			}
			return true;
		}
	}
}
